types/mempool/sender_nonce.go

```diff
@@ -0,0 +1,207 @@
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
+
+	sig := sigs[0]
+	sender := sig.PubKey.Address().String()
+	nonce := sig.Sequence
+	senderTxs, found := snm.senders[sender]
+	if !found {
+		senderTxs = huandu.New(huandu.Uint64)
+		snm.senders[sender] = senderTxs
+	}
+	senderTxs.Set(nonce, tx)
+
+	return nil
+}
+
+// Select returns an iterator ordering transactions the mempool with the lowest nonce of a random selected sender first.
+func (snm *senderNonceMempool) Select(_ sdk.Context, _ [][]byte) Iterator {
+	var senders []string
+	senderCursors := make(map[string]*huandu.Element)
+
+	orderedSenders := huandu.New(huandu.String)
+
+	// #nosec
+	for s := range snm.senders {
+		orderedSenders.Set(s, s)
+	}
+
+	s := orderedSenders.Front()
+	for s != nil {
+		sender := s.Value.(string)
+		senders = append(senders, sender)
+		senderCursors[sender] = snm.senders[sender].Front()
+		s = s.Next()
+	}
+
+	iter := &senderNonceMepoolIterator{
+		senders:       senders,
+		rnd:           snm.rnd,
+		senderCursors: senderCursors,
+	}
+
+	newIter := iter.Next()
+	if newIter == nil {
+		return nil
+	}
+	return newIter
+}
+
+// CountTx returns the total count of txs in the mempool.
+func (snm *senderNonceMempool) CountTx() int {
+	count := 0
+
+	// Disable gosec here since we need neither strong randomness nor deterministic iteration.
+	// #nosec
+	for _, value := range snm.senders {
+		count += value.Len()
+	}
+	return count
+}
+
+// Remove removes a tx from the mempool. It returns an error if the tx does not have at least one signer or the tx
+// was not found in the pool.
+func (snm *senderNonceMempool) Remove(tx sdk.Tx) error {
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
+
+	sig := sigs[0]
+	sender := sig.PubKey.Address().String()
+	nonce := sig.Sequence
+	senderTxs, found := snm.senders[sender]
+	if !found {
+		return ErrTxNotFound
+	}
+
+	res := senderTxs.Remove(nonce)
+	if res == nil {
+		return ErrTxNotFound
+	}
+
+	if senderTxs.Len() == 0 {
+		delete(snm.senders, sender)
+	}
+	return nil
+}
+
+type senderNonceMepoolIterator struct {
+	rnd           *rand.Rand
+	currentTx     *huandu.Element
+	senders       []string
+	senderCursors map[string]*huandu.Element
+}
+
+// Next returns the next iterator state which will contain a tx with the next smallest nonce of a randomly
+// selected sender.
+func (i *senderNonceMepoolIterator) Next() Iterator {
+	for len(i.senders) > 0 {
+		senderIndex := i.rnd.Intn(len(i.senders))
+		sender := i.senders[senderIndex]
+		senderCursor, found := i.senderCursors[sender]
+		if !found {
+			i.senders = removeAtIndex(i.senders, senderIndex)
+			continue
+		}
+
+		if senderCursor == nil {
+			i.senders = removeAtIndex(i.senders, senderIndex)
+			continue
+		}
```

Comment:
Is it possible for `senderCursor` to be nil if we've checked for `found`?

Reply:
Yes, `senderCursor.Next()` returns `nil` at the end of the list, and this is the stop condition.   We could also preemptively check and `removeAtIndex` below.  What do you think @JeancarloBarrios ?